Save bills

JonathanTylerWeber opened the pull request:
**Issue**
Create functionality for saving/unsaving bills and viewing saved bills on the dashboard

**Solution/Implementation Details**
Created a new saved bill model with prisma and migrated. Created a save button client component that will check to see if the bill is saved and handle saving/unsaving. Added api routes for saving, unsaving, checking if saved, and fetching saved bills. Updated dashboard page to display saved bills.

**Testing**
Go to a bill page and save, then go to dashboard to view. On the dashboard you can unsave it and on refresh it won't be on the page again. (Maybe could remove it from the dashboard immediately but I figure in case you accidentally click unsave and it immediately disappears, it may be hard to navigate to that specific bill again).

**Additional Information**
As for the migration stuff we discussed (adding session id to the model) I referenced this:
https://www.prisma.io/docs/orm/prisma-migrate/workflows/customizing-migrations

I needed to add the session id to go to the bill page from the dashboard. (This works as is without the session id, but the session id in the url is undefined when navigating to the bill page from the dashboard)

JonathanTylerWeber commented:
Probably should also remove all of the console.logs.

SevanBadal commented on app/api/auth/[...nextauth]/route.ts:
Needed to separate our config from the route to satisfy build requirements.

SevanBadal commented on app/api/saveBill/route.ts:
added sessionId to the object param of a created bill

SevanBadal commented on app/api/unsaveBill/route.ts:
This is not enough protection to ensure the user making the request is the owner of the bill being deleted. In this situation, any authenticated user can delete any bill even if they don't own the bill. We should first fetch the bill using the Prisma Client, check that the user on the bill is the session user, then delete the bill after ensuring they are the owner.

SevanBadal commented on app/dashboard/page.tsx:
```diff
@@ -1,5 +1,63 @@
+                                />
+                                <Link
+                                    className=" font-semibold leading-6 text-blue-600"
+                                    href={`/state/${bill.state}/session/${bill.sessionId}/bill/${bill.legiscanBillId}/history`}>
```
this was referencing bill.session which doesn't exist. I changed this to bill.sessionId.

SevanBadal commented on next-auth.d.ts:
This is pretty cool! Can you link me to where you discovered this solution?

SevanBadal commented on prisma/migrations/20240817160529_add_session_id_without_default/migration.sql:
bills will always have a session id.

SevanBadal reviewed (changes requested):
I pushed a few commits to fix the schema and some logic around including session id when creating a saved bill and routing to the bill show page from saved bills. 

There is one additional change we need around saved bill deletion authorization - see my comment for details.

SevanBadal commented on app/state/[state]/session/[session]/bill/[bill]/saveButton.tsx:
```diff
@@ -0,0 +1,105 @@
+            'Content-Type': 'application/json',
+          },
+          body: JSON.stringify({
+            legiscanBillId: bill.bill_id,
```
Bug here. Resaving a billl in the dashboard after unsaving creates a server side error: 

```
Error saving bill: PrismaClientValidationError: 
Invalid `prisma.savedBill.create()` invocation:

{
  data: {
    userId: 1,
    title: "House resolution relating to House Rules",
    description: "House resolution relating to House Rules",
    state: "VT",
    sessionTitle: "Unknown Session",
    sessionId: 0,
    changeHash: undefined,
+   legiscanBillId: Int
  }
}

Argument `legiscanBillId` is missing.
    at An (/Users/zomb/code/playground/legi/legibill/node_modules/@prisma/client/runtime/library.js:114:7526)
    at _n.handleRequestError (/Users/zomb/code/playground/legi/legibill/node_modules/@prisma/client/runtime/library.js:121:7392)
    at _n.handleAndLogRequestError (/Users/zomb/code/playground/legi/legibill/node_modules/@prisma/client/runtime/library.js:121:7057)
    at _n.request (/Users/zomb/code/playground/legi/legibill/node_modules/@prisma/client/runtime/library.js:121:6741)
    at async l (/Users/zomb/code/playground/legi/legibill/node_modules/@prisma/client/runtime/library.js:130:9355)
    at async POST (webpack-internal:///(rsc)/./app/api/saveBill/route.ts:26:27)
    at async /Users/zomb/code/playground/legi/legibill/node_modules/next/dist/compiled/next-server/app-route.runtime.dev.js:6:62609 {
  clientVersion: '5.18.0'
}
```

The `legiscanBillId` property should be set with `legiscanBillId` value defined at the top of this component. When the bill is unsaved, it is still shaped as the bill returned back from the db so it doesn't have a `bill_id`.

JonathanTylerWeber replied:
This pr should resolve it. I had to ensure that it could receive both named data types from either legiscan or our db (ie bill_id or legiscanBillId, change_hash or changeHash).
I also found that clicking the button quickly in succession would save the same bill multiple times so I added an isProcessing state to prevent that.

JonathanTylerWeber commented:
Fixed the unsave and then save bug on the dashboard. I had to ensure that it could receive both named data types from either legiscan or our db (ie bill_id or legiscanBillId, change_hash or changeHash).
I also found that clicking the button quickly in succession would save the same bill multiple times so I added an isProcessing state to prevent that.

SevanBadal reviewed (approved):
Approving. I think we can do a lot to simplify the `saveButton` component i.e. clean up the `handleClick` event callback.
